### PR TITLE
Avoid allocations for CIDs

### PR DESCRIPTION
**Pull request by Matthias247**
When reading CIDs from packets the current code makes some short-lived
allocations due to the use of `copy_to_bytes`, which allocate for 1.5% of
CPU samples. This change avoids this.

**Before:**
```
Sent 1073741824 bytes on 1 streams in 1.72s (594.74 MiB/s)
```

**After:**
```
Sent 1073741824 bytes on 1 streams in 1.70s (602.78 MiB/s)
```

### CODE REVIEW

File: quinn-proto/src/shared.rs [hunk 1]
```diff
@@ -75,17 +75,31 @@ impl ConnectionId {
...
         res
     }
 
     /// Decode from long header format
     pub(crate) fn decode_long(buf: &mut impl Buf) -> Option<Self> {
         let len = buf.get::<u8>().ok()? as usize;
-        match len > MAX_CID_SIZE || buf.remaining() < len {
-            false => Some(ConnectionId::new(&buf.copy_to_bytes(len))),
-            true => None,
+        if len > MAX_CID_SIZE || buf.remaining() < len {
```
Comment (djc) [inline quinn-proto/src/shared.rs:98]: The style change here seems unrelated?

**Review by djc — COMMENTED** [review]
Seems nice!

File: quinn-proto/src/shared.rs (hunk 1 above)
Comment (Matthias247) [reply quinn-proto/src/shared.rs:98]: Kind of. There was first more code in the `Some` branch, and I moved it out then because I know you prefer early returns instead of big if/else blocks. Then I added the delegate but kept it. I can change it if you prefer

Comment (djc) [reply quinn-proto/src/shared.rs:98]: Makes sense. I do prefer if you change it back.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Thanks!
